### DISCUSSION TITLE
How to install gcc in Mac M1

### DISCUSSION THREAD

**Discussion by anurawat01** [Homebrew/discussions#708] [Getting started] [answered]
How can I install gcc using **brew install gcc**?
I have tried but I got this error?

```
Searching for similarly named formulae...
Error: No similarly named formulae found.
Error: No available formula or cask with the name "gcc".
==> Searching for a previously deleted formula (in the last month)...
Error: No previously deleted formula found.
==> Searching taps on GitHub...
Error: No formulae found in taps.
```

Can anyone help me?

**Comment by SMillerDev** [comment, marked as answer]
What does `brew config` and `brew doctor` say?

**Reply by anurawat01** [reply]
Actually I am noob. I dont know how does this work brew config and brew doctor?

**Reply by SMillerDev** [reply]
Run the command and paste what it says here.

**Reply by anurawat01** [reply]
Both are working fine!
Anything particular you want to know?

**Reply by carlocab** [reply]
You need to

>paste what it says here

**Reply by anurawat01** [reply]
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some taps are not on the default git origin branch and may not receive
updates. If this is a surprise to you, check out the default branch with:
  git -C $(brew --repo homebrew/core) checkout master
anubhav@Anubhavs-MacBook-Air ~ % brew config
HOMEBREW_VERSION: 2.7.7
ORIGIN: https://github.com/Homebrew/brew
HEAD: cd191a7f6c4c8e053336d0ea34a9886b12d06470
Last commit: 8 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: (none)
Core tap last commit: never
Core tap branch: HEAD
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 12.0 build 1200
Git: 2.24.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 11.0.1-arm64
CLT: 12.4.0.0.1.1610135815
Xcode: N/A
Rosetta 2: false

**Reply by carlocab** [reply]
This is your problem, I think:
```
Warning: Some taps are not on the default git origin branch and may not receive
updates. If this is a surprise to you, check out the default branch with:
git -C $(brew --repo homebrew/core) checkout master
```
Try that command, do `brew update`, then try to install `gcc` again.

**Reply by anurawat01** [reply]
Yeah I have tried that as well!

**Reply by carlocab** [reply]
Ok, what do `brew config` and `brew doctor` say after doing that?

**Reply by anurawat01** [reply]
```
HOMEBREW_VERSION: 3.0.0-13-g15ae53c
ORIGIN: https://github.com/Homebrew/brew
HEAD: 15ae53c049c498d9f07ddc9e208c1d7dd56857fd
Last commit: 38 minutes ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: (none)
Core tap last commit: never
Core tap branch: HEAD
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 12.0 build 1200
Git: 2.24.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 11.0.1-arm64
CLT: 12.4.0.0.1.1610135815
Xcode: N/A
Rosetta 2: false
```

**Reply by anurawat01** [reply]
Main error I am getting while install gcc is:
```
fatal: Could not resolve HEAD to a revision
```
I think problem is in head?

**Reply by carlocab** [reply]
What happens when you do
```
cd $(brew --repo homebrew/core)
git pull --ff-only
```
?

Paste the output here please.

**Reply by anurawat01** [reply]
```
There is no tracking information for the current branch.
Please specify which branch you want to merge with.
See git-pull(1) for details.

    git pull <remote> <branch>

If you wish to set tracking information for this branch you can do so with:

    git branch --set-upstream-to=origin/<branch> master
    
 ```

**Reply by carlocab** [reply]
Ok, still in the same directory (`cd` again if you have to), what is the output of
```
git remote -v
```
?

**Reply by anurawat01** [reply]
```
origin	https://github.com/Homebrew/homebrew-core (fetch)
origin	https://github.com/Homebrew/homebrew-core (push)
```

**Reply by carlocab** [reply]
Weird, ok. Try
```
git branch --set-upstream-to=origin/master master
git pull --ff-only
```

**Reply by anurawat01** [reply]
In same dir?

**Reply by carlocab** [reply]
Yes, any `git` commands need to be done in `$(brew --repo homebrew/core)`.

**Reply by anurawat01** [reply]
```
fatal: branch 'master' does not exist
There is no tracking information for the current branch.
Please specify which branch you want to merge with.
See git-pull(1) for details.

    git pull <remote> <branch>

If you wish to set tracking information for this branch you can do so with:

    git branch --set-upstream-to=origin/<branch> master

```

**Reply by carlocab** [reply]
What about

    git pull --ff-only origin master

?

**Reply by anurawat01** [reply]
```
From https://github.com/Homebrew/homebrew-core
 * branch                  master     -> FETCH_HEAD
 ```

**Reply by carlocab** [reply]
That might've worked. Can you paste the output of `brew config` again? Then try to do `brew install gcc`.

**Reply by anurawat01** [reply]
Yeah it working now.... Could you please tell what was the problem?

**Reply by gromgit** [reply]
Your core tap Git repo wasn't initialized correctly. Since you also started #698, I'm guessing that you never actually completed the install successfully.

**Reply by guptajiakshat** [reply]
HOMEBREW_VERSION: 3.0.5-42-ge3b451c
ORIGIN: https://github.com/Homebrew/brew
HEAD: e3b451cce8b555da5c4043767fe1659d14d04264
Last commit: 2 days ago
Core tap ORIGIN: https://github.com/Homebrew/brew
Core tap HEAD: e3b451cce8b555da5c4043767fe1659d14d04264
Core tap last commit: 2 days ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit kabylake
Clang: 12.0 build 1200
Git: 2.24.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 11.2.2-x86_64
CLT: 12.0.32.2
Xcode: N/A
akshatgupta@Akshats-MacBook-Pro ~ %

**Reply by guptajiakshat** [reply]
i need to run pbds on mac any kind of help will be appreciated

**Reply by gromgit** [reply]
>```
>Core tap ORIGIN: https://github.com/Homebrew/brew
>```

That is absolutely the wrong repo to be cloning as the core tap. Run `brew doctor` and fix everything it warns you about; this will be one of the things it'll complain about.

**Reply by iamharshilmodi** [reply]
Thanks, @carlocab I had the same problem as @anurawat01. It got solved now. cheers!

**Reply by amiitk2107** [reply]
@iamharshilmodi, Can you help me in solving that problem.
I am unable to solve that problem

**Reply by gromgit** [reply]
@amiitk2107 , run `brew doctor` and fix everything it tells you to fix, then see if you still have "that problem".

**Comment by 2sn** [comment]
Since recently I have an issue with gfortran - no longer being able to make executable
```
~/test>gfortran test9.f90
ld: library not found for -lm
collect2: error: ld returned 1 exit status
```
It used to work just a few days ago and I am not really sure what may have changed.  I use MacOS 12.0.1 on M1 (mac mini).

It will also not be finding `-lSystem`

I can't really locate these files (`libm.a` or `libm.dylib`) on the machine - hence have no idea whether the issue may be related to a poor library path.  I tried to install recent Xcode (3.2 beta) and reinstalled homebrew, but all to no avail.  I am not sue whether this is an issue with homebrew or otherwise.

Any help or thoughts would be much appreciated.

**Reply by carlocab** [reply]
System libraries no longer live in the file system since Big Sur. Using a beta version of Xcode seems like a recipe for trouble. What is the output of `brew config` and `brew doctor`?

**Comment by 2sn** [comment]
Well, I had installed Xcode 3.2 beta after things did not work with current stable Xcode that was installed.

**Comment by 2sn** [comment]
With some web search there was an indication to directly put the path to `libm.dylib` on the compiler command line.  Then it worked.  It seems in older versions of `homebrew/gcc` the path had been set correctly.  As far as I can tell, there was no update on MacOS between my last successful compilation and the failure, only change I recall is a `brew upgrade` installing a new version of `gcc`.

**Comment by 2sn** [comment]
The solution I took, hence is to add it in `.bashrc` to `LIBRARY_PATH`:
```
export LIBRARY_PATH=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/lib:$LIBRARY_PATH
```
with that it works irrespective of Xcode 3.1 or 3.2 beta

**Comment by 2sn** [comment]
Now it is broken again, `f2py` produces modules that no longer loads.  

```
ImportError: dlopen(/Users/alex/python/source/double/_interface.cpython-310-darwin.so, 0x0002): symbol not found in flat namespace '___binary_MOD_direct'
```

Seems `homebrew` is suffering some melt-down.

**Comment by 2sn** [comment]
I think I now see the issue in the outputs
```
ld: warning: ignoring file /Volumes/a/alex/python/source/double/_library/library.a, building for macOS-arm64 but attempting to link with file built for macOS-arm64
```
Seems like someone programming linkers has decided `1 != 1`

**Comment by 2sn** [comment]
OK, I found that if I use the new macOS archive tool at
```
/Applications/Xcode.app/Contents//Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ar
```
then the linking works.  So it seems to be an issue with homebrew binutils.

**Reply by carlocab** [reply]
Homebrew binutils is GNU binutils, and this is keg-only because it often breaks builds when you add it to your `PATH`.

**Reply by 2sn** [reply]
Thanks for the clarification.  I seem to be needing some of then.  Maybe it was mostly this, but some MacOS updates have not propagated into it, or something broke just with recent Xcode.  I just saw various things break over the past few days ... Apple moved around libraries and modified the linker.
